libraries/mediaviewer/impl/src/main/kotlin/io/element/android/libraries/mediaviewer/impl/floatingvideo/FloatingVideoService.kt

```diff
@@ -0,0 +1,305 @@
+import android.os.IBinder
+import android.os.Looper
+import android.provider.Settings
+import android.view.Gravity
+import android.view.View
+import android.view.WindowManager
+import android.widget.Toast
+import android.widget.VideoView
+import androidx.compose.ui.platform.ComposeView
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.ViewModelStore
+import androidx.lifecycle.ViewModelStoreOwner
+import androidx.lifecycle.setViewTreeLifecycleOwner
+import androidx.lifecycle.setViewTreeViewModelStoreOwner
+import io.element.android.libraries.mediaviewer.impl.viewer.MediaViewerPageData
+import timber.log.Timber
+import androidx.savedstate.SavedStateRegistry
+import androidx.savedstate.SavedStateRegistryController
+import androidx.savedstate.SavedStateRegistryOwner
+import androidx.savedstate.setViewTreeSavedStateRegistryOwner
+import androidx.core.net.toUri
+import io.element.android.libraries.mediaviewer.impl.floatingvideo.ui.FloatingVideoOverlay
+import io.element.android.libraries.mediaviewer.impl.floatingvideo.util.getScreenHeight
+import io.element.android.libraries.mediaviewer.impl.floatingvideo.util.getScreenWidth
+import dev.zacsweers.metro.Inject
+import io.element.android.libraries.architecture.bindings
+
+class FloatingVideoService : Service(), LifecycleOwner, ViewModelStoreOwner, SavedStateRegistryOwner {
+    private var windowManager: WindowManager? = null
+    private var floatingView: View? = null
+    private var videoView: VideoView? = null
+    private var currentVideoData: MediaViewerPageData.MediaViewerData? = null
+    private var currentPosition: Long = 0L
+    private var isMaximized = true
+
+    private lateinit var windowLayoutParams: WindowManager.LayoutParams
+
+    companion object {
+        const val ACTION_START_FLOATING = "START_FLOATING"
+        const val ACTION_STOP_FLOATING = "STOP_FLOATING"
+        const val ACTION_UPDATE_POSITION = "UPDATE_POSITION"
+        const val EXTRA_VIDEO_ID = "video_id"
+        const val EXTRA_POSITION = "position"
+
+        @SuppressLint("ObsoleteSdkInt")
+        fun startFloating(
+            context: Context, videoData: MediaViewerPageData.MediaViewerData, position: Long = 0L
+        ) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(context)) {
+
+                //the message needs to be added into commonStrings as notice for permission needed
+                Toast.makeText(context, "To show the floating video, please allow 'Display over other apps' permission.", Toast.LENGTH_LONG).show()
+
+                // Request overlay permission
+                val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION).apply {
+                    data = "package:${context.packageName}".toUri()
+                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+                }
+
+
+                context.startActivity(intent)
```

Comment:
This code should be moved to `SystemUtils.kt` and the exception for `ActivityNotFoundException` should be caught.